src/library/filescanner.c

```diff
@@ -389,37 +409,83 @@ process_deferred_playlists(void)
     }
 }
 
-/* Thread: scan */
 static void
-process_file(char *file, time_t mtime, off_t size, int type, int flags, int dir_id)
+process_regular_file(char *file, struct stat *sb, int type, int flags, int dir_id)
 {
-  bool is_bulkscan;
-  bool is_type_compilation;
-  enum media_kind media_kind;
-  enum data_kind data_kind;
+  time_t stamp;
+  int id;
+  bool is_bulkscan = (flags & F_SCAN_BULK);
+  struct media_file_info mfi;
+  char virtual_path[PATH_MAX];
+  int ret;
 
-  is_bulkscan = (flags & F_SCAN_BULK);
+  // Do not rescan metadata if file did not change since last scan
+  db_file_stamp_bypath(file, &stamp, &id);
+  if (stamp && (stamp >= sb->st_mtime))
+    {
+      db_file_ping(id);
+      return;
+    }
 
-  switch (file_type_get(file))
+  // File is new or modified - (re)scan metadata and update file in library
+  memset(&mfi, 0, sizeof(struct media_file_info));
+
+  mfi.id = id;
+  mfi.fname = strdup(filename_from_path(file));
+  mfi.path = strdup(file);
+
+  mfi.time_modified = sb->st_mtime;
+  mfi.file_size = sb->st_size;
+
+  snprintf(virtual_path, PATH_MAX, "/file:%s", file);
+  mfi.virtual_path = strdup(virtual_path);
+
+  mfi.directory_id = dir_id;
+
+  if (S_ISFIFO(sb->st_mode))
     {
-      case FILE_REGULAR:
-	is_type_compilation = type & F_SCAN_TYPE_COMPILATION;
-	if (type & F_SCAN_TYPE_AUDIOBOOK)
-	  media_kind = MEDIA_KIND_AUDIOBOOK;
-	else if (type & F_SCAN_TYPE_PODCAST)
-	  media_kind = MEDIA_KIND_PODCAST;
-	else
-	  media_kind = 0;
+      mfi.data_kind = DATA_KIND_PIPE;
+      mfi.type = strdup("wav");
+      mfi.codectype = strdup("wav");
+      mfi.description = strdup("PCM16 pipe");
+      mfi.media_kind = MEDIA_KIND_MUSIC;
+    }
+  else
+    {
+      mfi.data_kind = DATA_KIND_FILE;
 
-	if (F_SCAN_TYPE_PIPE & type)
-	  data_kind = DATA_KIND_PIPE;
-	else
-	  data_kind = DATA_KIND_FILE;
+      if (type & F_SCAN_TYPE_AUDIOBOOK)
```

Comment:
If it was the intention that an audiobook can also be a podcast, this woud have never worked. The media_kind is assigned either audiobook or podcast, there is and as far as i have followed the history never were a bit wise or-assignement to media_kind. The change to "else if" only makes it a bit clearer that the audiobook-flag takes precedence over the podcast-flag.